File: 124. Binary Tree Maximum Path Sum/Java/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int maxPathSum(TreeNode root) {
        class Utils {
            int maxValue = Integer.MIN_VALUE;

            int pathSum(TreeNode root) {
                if (root == null) {
                    return 0;
                }
                int left = Math.max(0, pathSum(root.left));
                int right = Math.max(0, pathSum(root.right));
                maxValue = Math.max(maxValue, left + right + root.val);
                return Math.max(left, right) + root.val;
            }
        }
        Utils utils = new Utils();
        utils.pathSum(root);
        return utils.maxValue;
    }
}

File: 124. Binary Tree Maximum Path Sum/Python/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution(object):
    def maxPathSum(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        def path_sum(root, result):
            if not root:
                return (0, result)
            left, result = path_sum(root.left, result)
            left = max(0, left)
            right, result = path_sum(root.right, result)
            right = max(0, right)
            result = max(result, left + right + root.val)
            return max(left, right) + root.val, result

        return path_sum(root, -sys.maxsize)[1]
